### PR TITLE
Avoid cascading diagnostic for unexpected override

### LINKED ISSUE #31437: This error need to be correct or improved in a better way ?
I have a codebase in which parent class contain a method called Print()

` public override void Print(object sender, PrintPageEventArgs e)
        {
        }`

Now if I don't have a PrintPageEventArgs's namespace referenced in my .cs file it says "No suitable method found to override". This trick my mind to think that the print method doesn't exist in the base class (It might derived from the class which don't have Print() method to override) or  I am doing something wrong. When I add the namespace the problem get fixed.

what I am proposing here is simply giving an error for "please fix the namespace for this "PrintPageEventArgs". The current error "nothing to override" make confusion. I think I derived it from a class which doesn't have a print method to override.

I might be wrong here, My belief is showing the error to fix namespace instead of saying "Nothing to override".

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Check if there are any errors in a parameter type of an overriding method before reporting ERR_OverrideNotExpected.

Fixes #31437

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 1]
```diff
@@ -608,7 +608,57 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                        }
+                        if(!suppressError)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_OverrideNotExpected, overridingMemberLocation, overridingMember);
+                        }
+
+                        bool IsOrContainsErrorType(TypeSymbol symbol)
+                        {
+                            if (symbol.IsErrorType())
+                                return true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: There are multiple cases that changed in this method. But only one test was updated.
Please add tests to cover the remaining changes.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 2]
```diff
@@ -608,7 +608,57 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                                }
+                            }
+                            
+                        }
+                        if(!suppressError)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_OverrideNotExpected, overridingMemberLocation, overridingMember);
+                        }
+
+                        bool IsOrContainsErrorType(TypeSymbol symbol)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: The local function should be moved to the end of the method and the first letter lower-cased, by our conventions. Thanks

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: The purpose of this function is to visit all the nested types. We have a method for that already: `VisitType`.
It should be possible to use it instead.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 3]
```diff
@@ -608,7 +608,57 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                            {
+                                if (IsOrContainsErrorType(parameterType.TypeSymbol))
+                                {
+                                    suppressError = true; // The parameter type must be fixed before the override can be found, so suppress error
+                                    break;
+                                }
+                            }
+                            
+                        }
+                        if(!suppressError)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: nit: formatting `if (`

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: > Is UseSiteErrorTests the correct place to add the tests? or somewhere else?

The "use site errors" term is normally used to refer to errors returned by Symbol.GetUseSiteDiagnostic API. It doesn't look like this scenario is about them. 

---
In reply to: [242983943](https://github.com/dotnet/roslyn/pull/31937#discussion_r242983943) [](ancestors = 242983943)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: I see a lot of tests about ```ErrorCode.ERR_OverrideNotExpected``` in src\Compilers\CSharp\Test\Semantic\Semantics\InheritanceBindingTests.cs

---
In reply to: [243014444](https://github.com/dotnet/roslyn/pull/31937#discussion_r243014444) [](ancestors = 243014444,242983943)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Looking at the test files that reference `ERR_OverrideNotExpected`, `InheritanceBindingTests` stands out. That seems a good place to add such tests.

Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Thanks. I will do so when I get a chance.

**Comment by CyrusNajmabadi** [comment]
Can the titel of this PR be updated to something more appropriate?  Thanks!

**Comment by jcouv** [comment]
Adjusted title

**Comment by YairHalberstadt** [comment]
I have reworked the PR based on peer review feedback.
I have also suppressed the error when the parameter type of an indexer cannot be found.
I have also suppressed ERR_CantChangeTypeOnOverride (CS1715) and ERR_CantChangeReturnTypeOnOverride ( CS0508) when the return type of a property, indexer, method, or event cannot be found.

**Comment by YairHalberstadt** [comment]
All the builds seem to have failed, but looking at the errors messages, and considering the files I changed, that looks like a mistake. Is there anyway to rerun them?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 4]
```diff
@@ -608,7 +608,28 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                        bool suppressError = false;
+                        if (overridingMemberIsMethod || overridingMember.IsIndexer())
+                        {
+                            var parameterTypes = overridingMemberIsMethod
+                                ? ((MethodSymbol) overridingMember).ParameterTypes
+                                : ((PropertySymbol) overridingMember).ParameterTypes;
+
+                            foreach (var parameterType in parameterTypes)
+                            {
+                                if (parameterType.TypeSymbol.VisitType((typeSymbol, unused1, unused2) => typeSymbol.IsErrorType(), (object)null) != null)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: This is the cause of the build error. A change was merged yesterday that prohibits some uses of `==` on `TypeSymbol`.
The solution here is to add a cast on the left-hand-side operand: if ((object)... != null)`

```
2018-12-20T21:40:19.8558529Z F:\vsagent\6\s\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol_ImplementationChecks.cs(620,37): error CS0619: 'TypeSymbol.operator !=(TypeSymbol, TypeSymbol)' is obsolete: 'Use 'TypeSymbol.Equals(TypeSymbol, TypeSymbol, TypeCompareKind)' method.' [F:\vsagent\6\s\src\Compilers\CSharp\Portable\Microsoft.CodeAnalysis.CSharp.csproj]
2018-12-20T21:40:19.8576187Z F:\vsagent\6\s\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol_ImplementationChecks.cs(749,37): error CS0619: 'TypeSymbol.operator ==(TypeSymbol, TypeSymbol)' is obsolete: 'Use 'TypeSymbol.Equals(TypeSymbol, TypeSymbol, TypeCompareKind)' method.' [F:\vsagent\6\s\src\Compilers\CSharp\Portable\Microsoft.CodeAnalysis.CSharp.csproj]
2018-12-20T21:40:19.8576490Z F:\vsagent\6\s\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol_ImplementationChecks.cs(797,37): error CS0619: 'TypeSymbol.operator ==(TypeSymbol, TypeSymbol)' is obsolete: 'Use 'TypeSymbol.Equals(TypeSymbol, TypeSymbol, TypeCompareKind)' method.' [F:\vsagent\6\s\src\Compilers\CSharp\Portable\Microsoft.CodeAnalysis.CSharp.csproj]
2018-12-20T21:40:19.8576729Z F:\vsagent\6\s\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol_ImplementationChecks.cs(831,37): error CS0619: 'TypeSymbol.operator ==(TypeSymbol, TypeSymbol)' is obsolete: 'Use 'TypeSymbol.Equals(TypeSymbol, TypeSymbol, TypeCompareKind)' method.' [F:\vsagent\6\s\src\Compilers\CSharp\Portable\Microsoft.CodeAnalysis.CSharp.csproj]
```

Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: Or presumably I could use `! ... is null`?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: That would work too (but requires extra parens)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs [hunk 5]
```diff
@@ -693,6 +693,302 @@ class Derived : Base, Interface
...
+    public override int this[Int? x] => 0;
+}
+
+class Outer<T>
+{
+    public class Inner<U>{}
+}
+";
+            CompileAndVerifyDiagnostics(text, new ErrorDescription[] {
+                new ErrorDescription {Code = (int) ErrorCode.ERR_SingleTypeNameNotFound, Line = 8, Column = 30},
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InheritanceBindingTests.cs]: Isn't this style obsolete? probably should use VerifyDiagnostics

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 6]
```diff
@@ -608,7 +608,28 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
 
                     if ((object)associatedPropertyOrEvent == null)
                     {
-                        diagnostics.Add(ErrorCode.ERR_OverrideNotExpected, overridingMemberLocation, overridingMember);
+                        bool suppressError = false;
+                        if (overridingMemberIsMethod || overridingMember.IsIndexer())
+                        {
+                            var parameterTypes = overridingMemberIsMethod
+                                ? ((MethodSymbol) overridingMember).ParameterTypes
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: >`(MethodSymbol) overridingMember`

space. think that's R# default.

**Comment by YairHalberstadt** [comment]
I have fixed the build, and updated my PR based on peer review feedback.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 7]
```diff
@@ -608,7 +608,28 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
...
+                        bool suppressError = false;
+                        if (overridingMemberIsMethod || overridingMember.IsIndexer())
+                        {
+                            var parameterTypes = overridingMemberIsMethod
+                                ? ((MethodSymbol)overridingMember).ParameterTypes
+                                : ((PropertySymbol)overridingMember).ParameterTypes;
+
+                            foreach (var parameterType in parameterTypes)
+                            {
+                                if ((object)parameterType.TypeSymbol.VisitType((typeSymbol, unused1, unused2) => typeSymbol.IsErrorType(), (object)null) != null)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs]: >(object)parameterType.TypeSymbol.VisitType((typeSymbol, unused1, unused2) => typeSymbol.IsErrorType(), (object)null) != null [](start = 36, length = 124)

Consider extracting this to a local function, we're doing it 4 times.

**Comment by 333fred** [comment]
Done review pass (commit 7). Looks good overall, just one comment on code reuse.

**Comment by YairHalberstadt** [comment]
I have now extracted a local function as requested

**Comment by 333fred** [comment]
Thanks @YairHalberstadt!

**Pull request merged**